### ISSUE TITLE
DbCommands from DbProviderServices are not explicitly disposed

### ISSUE THREAD

**Issue by cincuranet** [aspnet/EntityFramework6#639]
The commands (`DbCommands`) from `DbProviderServices.CreateDbCommandDefinition` are not explicitly disposed and hence only later finalizer will release those. This is a problem, because during `IDisposable.Dispose` one could easily do whatever is needed, but when in finalizer, half of the "world" is already disposed and nothing can be done.

This in particular is a [problem](http://tracker.firebirdsql.org/browse/DNET-852) for `EntityFramework.Firebird` (and `FirebirdSql.Data.FirebirdClient`) package because unless the `Dispose` is properly called, server holds statements "ready" and eating memory as more statements are executed (this is problem especially in tight loops). In finalizer there's no way to send packet to server to release those statement. With that only closing the connection will eventually release the statements, however that might not happen because of i.e. loop and connection pooling.

Given the `DbCommands` (as a base class) has `IDisposable` EF should call the `Dispose` explicitly.

Basically similar issue #580, although that one also struggles with `DbParameter`, where the `IDisposable` is not on the base class, while it is for `DbCommand`.

### Further technical details
EF version: 6.2.0
Database Provider: N/A
Operating system: W10
IDE: Visual Studio 2017 15.8

**Comment by dpsenner** [comment]
I spent the last two hours with investigating if entity framework really creates disposable `DbCommand` instances without disposing them. The outcome appears clear to me while someone who has a better understanding of the EF6 internals may come to another conclusion. I would like to share my insights anyway to speed up the progress on this issue.

I have found several places in `ObjectContext` where instances of `DbCommand` or `EntityCommand` are created but not properly disposed. `EntityCommandDefinition` creates instances of `DbCommand` in `PrepareEntityCommandBeforeExecution` and both saves the created instance in `EntityCommand` via `SetStoreProviderCommand` and returns the created instance. `PrepareEntityCommandBeforeExecution` is invoked by `ExecuteStoreCommands` and `ExecuteStoreCommands` and the `DbDataReader` after command execution is returned by that method, not giving the invoker of `ExecuteStoreCommand` a chance to dispose the created `DbCommand`. Disposing the command in the method would be too early as there's still data to be read.

**Comment by ajcvickers** [comment]
@cincuranet @dpsenner We looked into this several years ago and even wrote code to dispose the commands, but the problem is that the EF stack has never disposed DbCommands and starting to do so raised questions about whether it was required, and whether or not they could be disposed before the reader and/or connection was disposed. At that time, we were not able to get anything working that didn't result in a breaking change on some provider. 

The group historical knowledge about why DbCommands aren't disposed by EF is:
* SqlCommand (for SQL Server) has never needed to be disposed
* DbCommand was only disposable because it inherited IDisposable from ComponentModel, for which it was there for a different reason.
* So either it was supposed to be the design that it didn't need to be disposed, or this was just missed because SQL Server didn't need it.

Based on all of this we decided that we will leave the code as-is and not dispose commands. This is the same behavior that has been in the code for over ten years now and, as far as we were aware, all providers that require disposing the command had already implemented workarounds. This usually involves disposing any open commands when the connection is disposed, or something similar. While it is conceivable hat we would take a change to fix this, we would have to be very sure that it didn't regress anything, and that is something that is hard to be sure about in this case, so I think this behavior is unlikely to change in EF6.

**Comment by ajcvickers** [comment]
**EF Team Triage:** This issue is not something that our team is planning to address in the EF6.x code base.  This does not mean that we would not consider a community contribution to address this issue.

Moving forwards, our team will be fixing bugs, implementing small improvements, and accepting community contributions to the EF6.x code base. Larger feature work and innovation will happen in the EF Core code base (https://github.com/aspnet/EntityFramework).

**Closing an issue in the EF6.x project does not exclude us addressing it in EF Core. In fact, a number of popular feature requests for EF have already been implemented in EF Core (alternate keys, batching in SaveChanges, etc.).**

*BTW this is a canned response and may have info or details that do not directly apply to this particular issue. While we'd like to spend the time to uniquely address every incoming issue, we get a lot traffic on the EF projects and that is not practical. To ensure we maximize the time we have to work on fixing bugs, implementing new features, etc. we use canned responses for common triage decisions.*

**Comment by dpsenner** [comment]
@ajcvickers, I understand the reasoning. Thanks for the insights. Note that Oracle connectors suffer under the same issue (#580) with additional intrications. Is this an official statement that ef6 is in maintenance mode and users should migrate to ef core?

**Comment by ajcvickers** [comment]
@dpsenner I don't know what constitutes an "official statement" but this is the same messaging as we have been using for a few years now.

**Comment by cincuranet** [comment]
@ajcvickers Yes, I implemented workaround as well about 10 years ago. But recently as I fixed some finalizers the issue surfaced again, so I gave it a try here. :)

On the other hand, it's not like this needs to be a breaking change. Previously, as new features were introduced `DbProviderManifest` had "flags" (`SupportsXxx` and so on) to see what providers implement. This might be easy and clean way to do it (see, I'm not giving up and next time I'm in the office I'll bring it up again ;)).
